### ISSUE TITLE
Threading crash with closures

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7523]
Migrated from [rt.perl.org#31851](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31851) (status was 'resolved')

Searchable as RT31851$

**Comment by p5pRT** [comment]
### From jgmyers@proofpoint.com

#### Created by jgmyers@proofpoint.com



The following test program crashes on a multiprocessor box\.  Wrapping
the closure inside of an eval eliminates the crash&#8203;:

&nbsp;        my $obj = save\(eval 'sub \{ $bar = shift; \}'\);

The various threads all share one underlying CV object; the object is
not marked permanent\.  The CV's refcount is modified without locking\,
resulting in the CV being prematurely freed\.

use threads;

sub init \{
&nbsp;    my $thr;

&nbsp;    for \(my $i = 0; $i \< 8; $i\+\+\) \{
&nbsp;        $thr = threads\->new\(\\&task\);
&nbsp;    \}

&nbsp;    foreach $thr \(threads\->list\) \{
&nbsp;        if \($thr\->tid && \!threads&#8203;::equal\($thr\, threads\->self\)\) \{
&nbsp;            $thr\->join;
&nbsp;        \}
&nbsp;    \}
\}

sub save
\{
&nbsp;    my $self = \{\};
&nbsp;    $self\->\{arg\} = shift;
&nbsp;    return $self;
\}

sub task \{
&nbsp;    print "task starting\\n";
&nbsp;    for \(my $i = 1; $i \< 10000; $i\+\+\) \{
&nbsp;        my $bar;
&nbsp;        my $obj = save\(sub \{ $bar = shift; \}\);
&nbsp;    \}
\}

init\(\);

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.5:

Configured by jgmyers at Tue Sep 28 10:04:28 PDT 2004.

Summary of my perl5 (revision 5 version 8 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.4.21-20.elsmp, archname=i686-linux-thread-multi
    uname='linux pong 2.4.21-20.elsmp #1 smp wed aug 18 20:46:40 edt 
2004 i686 i686 i386 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define 
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS 
-DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include 
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING 
-fno-strict-aliasing -pipe -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.3.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E 
-Wl,-rpath,/u/jgmyers/perl/lib/5.8.5/i686-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
   


@INC for perl v5.8.5:
    /u/jgmyers/perl/lib/5.8.5/i686-linux-thread-multi
    /u/jgmyers/perl/lib/5.8.5
    /u/jgmyers/perl/lib/site_perl/5.8.5/i686-linux-thread-multi
    /u/jgmyers/perl/lib/site_perl/5.8.5
    /u/jgmyers/perl/lib/site_perl/5.8.3/i686-linux-thread-multi
    /u/jgmyers/perl/lib/site_perl/5.8.3
    /u/jgmyers/perl/lib/site_perl
    .


Environment for perl v5.8.5:
    HOME=/u/jgmyers
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=/tools/x/bin:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/u/jgmyers/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Oct 05\, 2004 at 05&#8203;:25&#8203;:15PM \-0000\, John Gardiner Myers wrote&#8203;:
> The following test program crashes on a multiprocessor box\.  Wrapping
> the closure inside of an eval eliminates the crash&#8203;:
> 
>         my $obj = save\(eval 'sub \{ $bar = shift; \}'\);
> 
> The various threads all share one underlying CV object; the object is
> not marked permanent\.  The CV's refcount is modified without locking\,
> resulting in the CV being prematurely freed\.

Thanks for the report\. Your analysis is \*nearly\* spot on ;\-\)
The threads don't share the CVs; a copy is cloned for each thread
that is created; however\, all the CV structs share the same OP tree\,
which is supposed to the read\-only\. The code that updates the refcount of
the op tree doesn't appear to be thread\-safe\.

op\_free\(\) does a OP\_REFCNT\_LOCK before decrementing it\, but none of
the code that increments it appears to do so\.

I'll try to look into this properly when I get more time\.

\(P5Pers&#8203;: is there any reason not to just wrap the OpREFCNT\_inc/OpREFCNT\_dec
macrcos with OP\_REFCNT\_LOCK/OP\_REFCNT\_UNLOCK ?\)

Dave\.

\-\- 
You live and learn \(although usually you just live\)\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Dave Mitchell wrote&#8203;:
> 
> op\_free\(\) does a OP\_REFCNT\_LOCK before decrementing it\, but none of
> the code that increments it appears to do so\.
> 
> I'll try to look into this properly when I get more time\.
> 
> \(P5Pers&#8203;: is there any reason not to just wrap the OpREFCNT\_inc/OpREFCNT\_dec
> macrcos with OP\_REFCNT\_LOCK/OP\_REFCNT\_UNLOCK ?\)

Seems sensible\.

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Oct 07\, 2004 at 12&#8203;:16&#8203;:55AM \+0100\, Dave Mitchell wrote&#8203;:
> Thanks for the report\. Your analysis is \*nearly\* spot on ;\-\)
> The threads don't share the CVs; a copy is cloned for each thread
> that is created; however\, all the CV structs share the same OP tree\,
> which is supposed to the read\-only\. The code that updates the refcount of
> the op tree doesn't appear to be thread\-safe\.
> 
> op\_free\(\) does a OP\_REFCNT\_LOCK before decrementing it\, but none of
> the code that increments it appears to do so\.

Now fixed by the change below\.

It appears to make closure creation about 4% slower on ITHREAD builds\.

Dave\.

\-\- 
Nothing ventured\, nothing lost\.


Change 23433 by davem@&#8203;davem\-splatty on 2004/10/29 21&#8203;:04&#8203;:17

&nbsp;    \[perl \#31851\] Threading crash with closures
&nbsp;    various OpREFCNT\_inc\(\) operations weren't doing locking

Affected files \.\.\.

\.\.\. //depot/perl/pad\.c\#37 edit
\.\.\. //depot/perl/regcomp\.c\#340 edit
\.\.\. //depot/perl/sv\.c\#767 edit

Differences \.\.\.

==== //depot/perl/pad\.c\#37 \(text\) ====

@&#8203;@&#8203; \-1410\,7 \+1410\,9 @&#8203;@&#8203;
 \#endif
&nbsp;     CvGV\(cv\)        = CvGV\(proto\);
&nbsp;     CvSTASH\(cv\)        = CvSTASH\(proto\);
\+    OP\_REFCNT\_LOCK;
&nbsp;     CvROOT\(cv\)        = OpREFCNT\_inc\(CvROOT\(proto\)\);
\+    OP\_REFCNT\_UNLOCK;
&nbsp;     CvSTART\(cv\)        = CvSTART\(proto\);
&nbsp;     CvOUTSIDE\(cv\)    = \(CV\*\)SvREFCNT\_inc\(outside\);
&nbsp;     CvOUTSIDE\_SEQ\(cv\) = CvOUTSIDE\_SEQ\(proto\);

==== //depot/perl/regcomp\.c\#340 \(text\) ====

@&#8203;@&#8203; \-4983\,9 \+4983\,14 @&#8203;@&#8203;
&nbsp;             \(SvTYPE\(new\_comppad\) == SVt\_PVAV\) ?
&nbsp;                     new\_comppad : Null\(PAD \*\)
&nbsp;         \);
\+        OP\_REFCNT\_LOCK;
&nbsp;         if \(\!OpREFCNT\_dec\(\(OP\_4tree\*\)r\->data\->data\[n\]\)\) \{
\+            OP\_REFCNT\_UNLOCK;
&nbsp;                     op\_free\(\(OP\_4tree\*\)r\->data\->data\[n\]\);
&nbsp;         \}
\+        else \{
\+            OP\_REFCNT\_UNLOCK;
\+        \}
 
&nbsp;         PAD\_RESTORE\_LOCAL\(old\_comppad\);
&nbsp;         SvREFCNT\_dec\(\(SV\*\)new\_comppad\);

==== //depot/perl/sv\.c\#767 \(text\) ====

@&#8203;@&#8203; \-10244\,7 \+10244\,9 @&#8203;@&#8203;
&nbsp;         case 'o'&#8203;:
&nbsp;         /\* Compiled op trees are readonly\, and can thus be
&nbsp;            shared without duplication\. \*/
\+        OP\_REFCNT\_LOCK;
&nbsp;         d\->data\[i\] = \(void\*\)OpREFCNT\_inc\(\(OP\*\)r\->data\->data\[i\]\);
\+        OP\_REFCNT\_UNLOCK;
&nbsp;         break;
&nbsp;         case 'n'&#8203;:
&nbsp;         d\->data\[i\] = r\->data\->data\[i\];
@&#8203;@&#8203; \-10955\,7 \+10957\,9 @&#8203;@&#8203;
&nbsp;     Perl\_rvpv\_dup\(aTHX\_ dstr\, sstr\, param\);
&nbsp;     CvSTASH\(dstr\)    = hv\_dup\(CvSTASH\(sstr\)\, param\); /\* NOTE&#8203;: not refcounted \*/
&nbsp;     CvSTART\(dstr\)    = CvSTART\(sstr\);
\+    OP\_REFCNT\_LOCK;
&nbsp;     CvROOT\(dstr\)    = OpREFCNT\_inc\(CvROOT\(sstr\)\);
\+    OP\_REFCNT\_UNLOCK;
&nbsp;     CvXSUB\(dstr\)    = CvXSUB\(sstr\);
&nbsp;     CvXSUBANY\(dstr\)    = CvXSUBANY\(sstr\);
&nbsp;     if \(CvCONST\(sstr\)\) \{

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From ajsavige@yahoo.com.au

Further to John Myers' closure threading crash\, I think I've
stumbled across the same problem\, as further discussed here&#8203;:

\<http&#8203;://www\.perlmonks\.org/?node\_id=409473>

I'm pleased to report that Dave Mitchell's patch \(Change 23433 by
davem on 2004/10/29 21&#8203;:04&#8203;:17\) does indeed solve my crashing problem
\(on both Linux and Windows\)\. Notice that my crashes seem to occur
on multi\-cpu machines \_only\_\.

To shoehorn Dave's patch into Windows\, I crudely hacked globals\.c
and perlvars\.h \(in perl 5\.8\.5\) as follows&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- globals-orig.c      Thu Apr 17 06:43:29 2003
+++ globals.c   Tue Nov 23 15:58:26 2004
@@ -16,6 +16,10 @@
 #define PERL_IN_GLOBALS_C
 #include "perl.h"

+#if defined(USE_ITHREADS) && defined(WIN32)
+__declspec(dllexport) perl_mutex PL_op_mutex;
+#endif
+
 int
 Perl_fprintf_nocontext(PerlIO *stream, const char *format, ...)
 {
--- perlvars-orig.h     Tue Aug 12 23:13:58 2003
+++ perlvars.h  Tue Nov 23 16:03:09 2004
@@ -44,7 +44,7 @@
 PERLVAR(Gmalloc_mutex, perl_mutex)     /* Mutex for malloc */
 #endif

-#if defined(USE_ITHREADS)
+#if defined(USE_ITHREADS) && !(defined(WIN32) && defined(PERL_IN_GLOBALS_C))
 PERLVAR(Gop_mutex,     perl_mutex)     /* Mutex for op refcounting */
 #endif


```
</details>
Note that this is not a clean fix.
I understand that Steve Hay is currently working on a clean fix
for this bug \(in "Smoke \[5\.9\.2\] 23450" thread\)\.

/\-\\


Find local movie times and trailers on Yahoo\! Movies\.
http&#8203;://au\.movies\.yahoo\.com

**Comment by p5pRT** [comment]
### From @steve-m-hay

Andrew Savige wrote&#8203;:

>Further to John Myers' closure threading crash\, I think I've
>stumbled across the same problem\, as further discussed here&#8203;:
>
>\<http&#8203;://www\.perlmonks\.org/?node\_id=409473>
>
>I'm pleased to report that Dave Mitchell's patch \(Change 23433 by
>davem on 2004/10/29 21&#8203;:04&#8203;:17\) does indeed solve my crashing problem
>\(on both Linux and Windows\)\. Notice that my crashes seem to occur
>on multi\-cpu machines \_only\_\.
>
>To shoehorn Dave's patch into Windows\, I crudely hacked globals\.c
>and perlvars\.h \(in perl 5\.8\.5\) as follows&#8203;:
>\[snip\]
>Note that this is not a clean fix\.
>I understand that Steve Hay is currently working on a clean fix
>for this bug \(in "Smoke \[5\.9\.2\] 23450" thread\)\.
>
Dave's patch should have been fixed for Win32 by change 23499&#8203;:

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse?patch=23499

\- Steve




---
Radan Computational Ltd\.

The information contained in this message and any files transmitted with it are confidential and intended for the addressee\(s\) only\.  If you have received this message in error or there are any problems\, please notify the sender immediately\.  The unauthorized use\, disclosure\, copying or alteration of this message is strictly forbidden\.  Note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of Radan Computational Ltd\.  The recipient\(s\) of this message should check it and any attached files for viruses&#8203;: Radan Computational will accept no liability for any damage caused by any virus transmitted by this email\.
